orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/monitoreddeploy/EvaluateDeploymentHealthTaskSpec.groovy

```diff
@@ -0,0 +1,145 @@
+      evaluateHealth(_) >> {
+        throw RetrofitError.networkError("url", new IOException())
+      }
+    }
+
+    def serviceProviderStub = Stub(DeploymentMonitorServiceProvider) {
+      getDefinitionById(_) >> {
+        def deploymentMonitor = new DeploymentMonitorDefinition()
+        deploymentMonitor.id = "LogMonitorId"
+        deploymentMonitor.name = "LogMonitor"
+        deploymentMonitor.failOnError = true
+        deploymentMonitor.service = monitorServiceStub
+
+        return deploymentMonitor
+      }
+
+    }
+
+    def task = new EvaluateDeploymentHealthTask(serviceProviderStub, new NoopRegistry())
+
+    MonitoredDeployStageData stageData = new MonitoredDeployStageData()
+    stageData.deploymentMonitor = new DeploymentMonitor()
+    stageData.deploymentMonitor.id = "LogMonitorId"
+    stageData.application = pipe.application
+
+    def stage = new Stage(pipe, "evaluateDeploymentHealth", mapper.convertValue(stageData, Map))
+    stage.startTime = Instant.now().toEpochMilli()
+
+    when:
+    TaskResult result = task.execute(stage)
+
+    then:
+    result.status == ExecutionStatus.RUNNING
+    result.context.deployMonitorHttpRetryCount == 1
+
+    when:
+    stage.context.deployMonitorHttpRetryCount = 3
+    result = task.execute(stage)
+
+    then:
+    result.status == ExecutionStatus.TERMINAL
+  }
+
+  def "should handle bad responses from 3rd party monitor"() {
+    given:
+    def monitorServiceStub = Stub(DeploymentMonitorService) {
+//    broadcastStarting(_) >> {
+//    }
+//
+//    broadcastCompleted(_) >> {
+//    }
+//
+      evaluateHealth(_) >> {
+        return new EvaluateHealthResponse()
+      } >> {
+        EvaluateHealthResponse response = new EvaluateHealthResponse()
+        response.nextStep = new DeploymentStep()
+
+        return response
+      }
+    }
+
+    def serviceProviderStub = Stub(DeploymentMonitorServiceProvider) {
+      getDefinitionById(_) >> {
+        def deploymentMonitor = new DeploymentMonitorDefinition()
+        deploymentMonitor.id = "LogMonitorId"
+        deploymentMonitor.name = "LogMonitor"
+        deploymentMonitor.failOnError = true
+        deploymentMonitor.service = monitorServiceStub
+
+        return deploymentMonitor
+      }
+    }
+
+    def task = new EvaluateDeploymentHealthTask(serviceProviderStub, new NoopRegistry())
+
+    MonitoredDeployStageData stageData = new MonitoredDeployStageData()
+    stageData.deploymentMonitor = new DeploymentMonitor()
+    stageData.deploymentMonitor.id = "LogMonitorId"
+    stageData.application = pipe.application
+
+    def stage = new Stage(pipe, "evaluateDeploymentHealth", mapper.convertValue(stageData, Map))
+    stage.startTime = Instant.now().toEpochMilli()
+
+    when:
+    TaskResult result = task.execute(stage)
+
+    then:
+    result.status == ExecutionStatus.TERMINAL
+    result.context.deploymentMonitorReasons == Collections.emptyList()
+
+    when:
+    result = task.execute(stage)
```

Comment:
is it the same when..then twice?

Reply:
it's sort of like defining "sub-tests", as best as I understand it. 'cuz `then` is not supposed to have sideeffects, right?

Reply:
added `description` for the `when`s